Add basic tests for RandomSpec

antoniojimeneznieto opened the pull request:
This PR aims to enhance the test coverage for the Random functionality by providing some initial test cases. As a first-time contributor, I apologize in advance for any mistakes and welcome any feedback. Here are some key points about the changes in this PR:

- The existing code has not been modified, ensuring that the changes in this PR are solely focused on adding tests.
- The tests for randomness algorithms like shuffle have been implemented with a low probability of failure (1/10000!). Although this approach might not be ideal for testing the randomness of these functions, it serves as a starting point for further discussion and improvement.

Closes https://github.com/typelevel/cats-effect/issues/2909.

antoniojimeneznieto committed:
Add tests for Random class

djspiewak commented:
This looks like a great start! Random test failure rates do concern me though. Even 0.01% is still comparatively frequent given the number of test runs we do on a regular basis, and random CI failures are the *worst*. I'm definitely not one to put up my nose at more tests though. Is there a way we can simply detect the collision case and retry, reducing the odds to something like 1 - (1 - 0.0001)^k, where *k* is the maximum number of retries?

antoniojimeneznieto commented:
Thank you for the feedback, Daniel. I appreciate it :) It seems that I may have misunderstood the probability of failure in this particular test. Given a list composed of numbers between 1 and 10,000, where each number is different and the order matters, there are 10,000! (factorial of 10,000) possible combinations. Assuming the shuffle is truly random, the probability of obtaining the exact same list is 1 / factorial(10,000), resulting in a failure probability of approximately 2.846 x 10^(-35659) (I may be missing something in my reasoning). Concerning your idea of detecting collisions and retrying sounds like an excellent alternative to me

antoniojimeneznieto committed:
Formatting

armanbilge commented on tests/shared/src/test/scala/cats/effect/std/RandomSpec.scala:
```diff
@@ -17,38 +17,238 @@
+        val numIterations: Int = 1000
+        for {
+          random <- Random.scalaUtilRandom[IO]
+          randDoubles <- (1 to numIterations)
+            .toList
+            .traverse(_ => random.betweenDouble(min, max))
```
```suggestion
          randDoubles <- random.betweenDouble(min, max).replicateA(numIterations)
```

antoniojimeneznieto replied:
Cleaner, more elegant and efficient. I am going to update the tests accordingly. Thank you Arman for the feedback :)

armanbilge commented on tests/shared/src/test/scala/cats/effect/std/RandomSpec.scala:
```diff
@@ -17,38 +17,238 @@
+      "generate random int value" in real {
+        for {
+          random <- Random.scalaUtilRandom[IO]
+          int <- random.nextInt
+        } yield int match {
+          case _: Int => true
+          case _ => false
+        }
```
Is this really asserting something that is not guaranteed by the type system? :)

A no-op assert is fine for this. Just executing `nextInt` asserts that it links and runs on Scala.js and Native which is always useful.

```suggestion
          int <- random.nextInt
        } yield ok
```

antoniojimeneznieto replied:
I see, it makes completely sense. I am going update it!

antoniojimeneznieto committed:
Refactor tests to use replicateA and no-op assert

Refactor code to use replicateA for a more cleaner and efficient type generation.

Refactor code to use no-op assert for confirming that nextInt, nextLong and nextBoolean links and runs on Scala.js and Native.

armanbilge commented:
Fatal warnings in CI :)

https://github.com/typelevel/cats-effect/actions/runs/4599225154/jobs/8124359599?pr=3516#step:23:148

```
[error] /home/runner/work/cats-effect/cats-effect/tests/shared/src/test/scala/cats/effect/std/RandomSpec.scala:91:11: parameter value randomBoolean in anonymous function is never used
[error]           randomBoolean <- random.nextBoolean
[error]           ^
[error] /home/runner/work/cats-effect/cats-effect/tests/shared/src/test/scala/cats/effect/std/RandomSpec.scala:162:11: parameter value int in anonymous function is never used
[error]           int <- random.nextInt
[error]           ^
[error] /home/runner/work/cats-effect/cats-effect/tests/shared/src/test/scala/cats/effect/std/RandomSpec.scala:182:11: parameter value long in anonymous function is never used
[error]           long <- random.nextLong
[error]           ^
[error] No warnings can be incurred under -Werror.
[error] four errors found
```

antoniojimeneznieto commented:
It would be a good idea to use must beTrue ? Or if the tests are not adding much value, deleting them could be another option ?

armanbilge commented:
Sorry, not sure I understood your question. But even if the tests are not asserting anything very interesting, they are still useful just for testing that the code can link and run.

antoniojimeneznieto commented:
Sorry, I should have explained it better, the problem seems to be that randomBoolean, int and long are not being utilised. Maybe instead of just yielding ok we could use the following matchers:

randomBoolean must beOneOf(true, false)
int must beBetween(minVal, maxVal)
long must beBetween(minVal, maxVal)

antoniojimeneznieto committed:
Avoid compiler error using randomBoolean, int, long

It appears that the variables randomBoolean, int, and long are not being utilized in the tests, which is generating a critical warning.

armanbilge commented:
Oh, I see!

You can assign the value to `_` to indicate that it is unused, we are only running it for the effect. e.g.
```
_ <- random.nextBoolean
```

armanbilge commented on tests/shared/src/test/scala/cats/effect/std/RandomSpec.scala:
```diff
@@ -89,7 +89,7 @@ class RandomSpec extends BaseSpec {
           random <- Random.scalaUtilRandom[IO]
           randomBoolean <- random.nextBoolean
-        } yield ok
+        } yield randomBoolean must beOneOf(true, false)
```
This is fine too :)

armanbilge commented on tests/shared/src/test/scala/cats/effect/std/RandomSpec.scala:
```diff
@@ -17,38 +17,228 @@
 package cats.effect
 package std
 
+import cats.implicits._
```
Nit: `cats.implicits` import is soft-deprecated, the syntax import can fully replace it :)
```suggestion
import cats.syntax.all._
```

antoniojimeneznieto replied:
Noted :)

armanbilge commented on tests/shared/src/test/scala/cats/effect/std/RandomSpec.scala:
```diff
@@ -17,38 +17,228 @@
+        val max: Double = 1.0
+        val numIterations: Int = 1000
+        for {
+          random <- Random.scalaUtilRandom[IO]
```
What do you think about abstracting these tests over the `Random` implementation? E.g. moving them inside of a `def(random: IO[Random[IO]])`.

Then we can run all the tests with each of the different random implementations (scala util, thread local, secure, etc.) since this is where we've found platform-specific bugs before (e.g. secure was not implemented on JS). I think we've also had bugs with the thread local random before as well, using the wrong thread local ... 😅

antoniojimeneznieto replied:
That's a brilliant idea, I am going to implement it :)

antoniojimeneznieto replied:
There are still a couple of methods to be abstracted, but what do you think so far?

armanbilge replied:
Thanks! Sorry, I should have given an example. Here's how we are doing it for `Dispatcher`:
https://github.com/typelevel/cats-effect/blob/397ddd109b0463955191789892aa8cc9b886fa7c/tests/shared/src/test/scala/cats/effect/std/DispatcherSpec.scala#L220-L222

antoniojimeneznieto replied:
My bad, I should have checked dispatcher before

armanbilge reviewed (commented):
Looking very good!

I had one last idea about testing the different `Random` implementations—that way we can maximize the value of your work :)

antoniojimeneznieto committed:
Abstract check in range and shuffle Sequence

antoniojimeneznieto committed:
Abstract tests

antoniojimeneznieto commented:
Could we reutilise the tests for SecureRandom? 

`"SecureRandom" should {
  testRandom(SecureRandom.javaSecuritySecureRandom[IO])
}`

armanbilge commented on tests/shared/src/test/scala/cats/effect/std/RandomSpec.scala:
```diff
@@ -22,39 +22,262 @@ import org.specs2.specification.core.Fragments
+
+    testRandom(R)
+
+    elementOfTests[Int, List[Int]](
```
Nice job! Is it possible to move these tests inside of `testRandom` as well?

Then as you point out, we should run these tests for all the different random constructors we have: `scalaUtilRandom`, `javaUtilRandom`, `javaUtilConcurrentThreadLocalRandom`, and `javaSecuritySecureRandom` I think are all of them

antoniojimeneznieto replied:
Thanks :) Yes!

antoniojimeneznieto committed:
TestRandom with different Randoms

djspiewak reviewed (approved):
This is great! Thank you!
